common.mk

```diff
@@ -111,42 +109,100 @@ $(FIRRTL_FILE) $(ANNO_FILE) &: $(SCALA_SOURCES) $(sim_files) $(SCALA_BUILDTOOL_D
 
+define firtool_extra_anno_contents
+[
+	{
+		"class":"sifive.enterprise.firrtl.MarkDUTAnnotation",
+		"target":"~$(MODEL)|$(TOP)"
+	},
+	{
+		"class": "sifive.enterprise.firrtl.TestHarnessHierarchyAnnotation",
+		"filename": "$(FIRTOOL_MODEL_HRCHY_JSON)"
+	},
+	{
+		"class": "sifive.enterprise.firrtl.ModuleHierarchyAnnotation",
+		"filename": "$(FIRTOOL_TOP_HRCHY_JSON)"
+	}
+]
+endef
+export firtool_extra_anno_contents
+$(FINAL_ANNO_FILE) $(FIRTOOL_EXTRA_ANNO_FILE): $(ANNO_FILE)
+	echo "$$firtool_extra_anno_contents" > $(FIRTOOL_EXTRA_ANNO_FILE)
+	jq -s '[.[][]]' $(ANNO_FILE) $(FIRTOOL_EXTRA_ANNO_FILE) > $(FINAL_ANNO_FILE)
+
 .PHONY: firrtl
-firrtl: $(FIRRTL_FILE)
+firrtl: $(FIRRTL_FILE) $(FINAL_ANNO_FILE)
 
 #########################################################################################
 # create verilog files rules and variables
 #########################################################################################
-REPL_SEQ_MEM = --infer-rw --repl-seq-mem -c:$(MODEL):-o:$(TOP_SMEMS_CONF)
-HARNESS_CONF_FLAGS = -thconf $(HARNESS_SMEMS_CONF)
-
-TOP_TARGETS = $(TOP_FILE) $(TOP_SMEMS_CONF) $(TOP_ANNO) $(TOP_FIR) $(sim_top_blackboxes)
-HARNESS_TARGETS = $(HARNESS_FILE) $(HARNESS_SMEMS_CONF) $(HARNESS_ANNO) $(HARNESS_FIR) $(sim_harness_blackboxes)
+FIRTOOL_TARGETS = \
+	$(FIRTOOL_SMEMS_CONF) \
+	$(FIRTOOL_TOP_SMEMS_JSON) \
+	$(FIRTOOL_TOP_HRCHY_JSON) \
+	$(FIRTOOL_MODEL_MOD_HRCHY_JSON) \
+	$(FIRTOOL_MODEL_SMEMS_JSON) \
+	$(FIRTOOL_FILELIST) \
+	$(FIRTOOL_BB_MODS_FILELIST)
 
 # DOC include start: FirrtlCompiler
-$(TOP_TARGETS) $(HARNESS_TARGETS) &: $(FIRRTL_FILE) $(ANNO_FILE) $(VLOG_SOURCES)
-	$(call run_scala_main,tapeout,barstools.tapeout.transforms.GenerateTopAndHarness,\
-		--allow-unrecognized-annotations \
-		--output-file $(TOP_FILE) \
-		--harness-o $(HARNESS_FILE) \
+# hack: lower to middle firrtl if Fixed types are found
+$(FIRTOOL_TARGETS) &: $(FIRRTL_FILE) $(FINAL_ANNO_FILE) $(VLOG_SOURCES)
+	$(call run_scala_main,tapeout,barstools.tapeout.transforms.GenerateTop,\
+		--no-dedup \
+		--output-file $(SFC_FIRRTL_BASENAME) \
+		--output-annotation-file $(SFC_ANNO_FILE) \
+		--target-dir $(OUT_DIR) \
 		--input-file $(FIRRTL_FILE) \
-		--syn-top $(TOP) \
-		--harness-top $(VLOG_MODEL) \
-		--annotation-file $(ANNO_FILE) \
-		--top-anno-out $(TOP_ANNO) \
-		--top-dotf-out $(sim_top_blackboxes) \
-		--top-fir $(TOP_FIR) \
-		--harness-anno-out $(HARNESS_ANNO) \
-		--harness-dotf-out $(sim_harness_blackboxes) \
-		--harness-fir $(HARNESS_FIR) \
-		$(REPL_SEQ_MEM) \
-		$(HARNESS_CONF_FLAGS) \
-		--target-dir $(build_dir) \
+		--annotation-file $(FINAL_ANNO_FILE) \
 		--log-level $(FIRRTL_LOGLEVEL) \
+		--allow-unrecognized-annotations \
+		-X $(if $(shell grep "Fixed<" $(FIRRTL_FILE)),middle,none) \
 		$(EXTRA_FIRRTL_OPTIONS))
-	touch $(sim_top_blackboxes) $(sim_harness_blackboxes)
+	$(if $(shell grep "Fixed<" $(FIRRTL_FILE)),mv $(SFC_FIRRTL_BASENAME).mid.fir $(SFC_FIRRTL_FILE),)
+	firtool \
+		--format=fir \
+		-O=release \
```

Comment:
We might also want to be able to toggle this flag between `-O=release` and `-O=debug`, since the latter could be useful for verification. @seldridge can probably provide more context on the intended use of these flags.

Reply:
You can also opt to just remove this as `-O=release` is the default. There's a lot of history to this option which is detailed in [this section](https://circt.llvm.org/docs/Dialects/FIRRTL/RationaleFIRRTL/#name-preservation-modes) of the FIRRTL Rational document (which is missing the final chapter, discussed below).

tl;dr: we are pretty confident that `-O=release` is now good enough for everyone as of the newest version of CIRCT, v1.19.0.

Having `-O=release` (preservation of named signals is best effort) and `-O=debug` (preservation of named signals is mandatory) as well as the underlying option that backs these `-preserve-values=[none, named, all]`) was a compromise between simulation speed/PPA and debuggability. Early versions of CIRCT were far too aggressive in name removal. However, after a ton of effort we've gotten it to a point where we believe (and have received feedback that) `-O=release` is better than the old SFC behavior (which was favored by designers) and better than existing MFC behavior of either `-O=release` or `-O=debug` to this point.

`-O=release` is already the default behavior and is expected to be the one, true optimization level that is used to compile designs. `-O=debug` may be useful if you are trying to debug something and desperately want to keep all named things from Chisel or `-preserve-values=all` may be an even stronger fallback of preserving named, temporary things from Chisel. However, neither `-O=debug` or `-preserve-values=all` are expected to be normal modes at which design or verification/PD use. For one-off debugging, I would suggest using a don't touch.